src/c1_state_machine/p5_digital_cash.rs

```diff
@@ -48,9 +99,197 @@ impl StateMachine for DigitalCashSystem {
+            spends: vec![
+                Bill { owner: User::Alice, amount: 40, serial: 0 },
+                Bill { owner: User::Alice, amount: 40, serial: 0 }
+            ],
+            receives: vec![
+                Bill { owner: User::Bob, amount: 20, serial: start.next_serial() },
+                Bill { owner: User::Bob, amount: 20, serial: start.next_serial() + 1 },
+                Bill { owner: User::Alice, amount: 40, serial: start.next_serial() + 2 }
+            ],
+        }
+    );
+    let expected = State::from([
+        Bill { owner: User::Alice, amount: 40, serial: 0 }
+    ]);
+    assert_eq!(end, expected);
+}
+
+#[test]
+fn spending_more_than_bill_fails() {
+    let start = State::from([
+        Bill { owner: User::Alice, amount: 40, serial: 0 },
+        Bill { owner: User::Charlie, amount: 42, serial: 1 }
+    ]);
+    let end = DigitalCashSystem::next_state(
+        &start,
+        &CashTransaction::Transfer {
+            spends: vec![
+                Bill { owner: User::Alice, amount: 40, serial: 0 },
+                Bill { owner: User::Charlie, amount: 42, serial: 1 }
+            ],
+            receives: vec![
+                Bill { owner: User::Bob, amount: 20, serial: 2 },
+                Bill { owner: User::Bob, amount: 20, serial: 3 },
+                Bill { owner: User::Alice, amount: 52, serial: 4 }
+            ],
+        }
+    );
+    let expected = State::from([
+        Bill { owner: User::Alice, amount: 40, serial: 0 },
+        Bill { owner: User::Charlie, amount: 42, serial: 1 }
+    ]);
+    assert_eq!(end, expected);
+}
+
+#[test]
+fn spending_non_existent_bill_fails() {
+    let start = State::from([
+        Bill { owner: User::Alice, amount: 32, serial: 0 },
+    ]);
+    let end = DigitalCashSystem::next_state(
+        &start,
+        &CashTransaction::Transfer {
+            spends: vec![
+                Bill { owner: User::Bob, amount: 1000, serial: 32 },
+            ],
+            receives: vec![
+                Bill { owner: User::Bob, amount: 1000, serial: 33 }
+            ]
+        }
+    );
+    let expected = State::from([
+        Bill { owner: User::Alice, amount: 32, serial: 0 },
+    ]);
+    assert_eq!(end, expected);
+}
+
+#[test]
+fn spending_from_alice_to_all() {
+    let start = State::from([
+        Bill { owner: User::Alice, amount: 42, serial: 0 }
+    ]);
+    let end = DigitalCashSystem::next_state(
+        &start,
+        &CashTransaction::Transfer {
+            spends: vec![
+                Bill { owner: User::Alice, amount: 42, serial: 0 }
+            ],
+            receives: vec![
+                Bill { owner: User::Alice, amount: 10, serial: 1 },
+                Bill { owner: User::Bob, amount: 10, serial: 2 },
+                Bill { owner: User::Charlie, amount: 10, serial: 3 }
+            ]
+        }
+    );
+    let expected = State::from([
+        Bill { owner: User::Alice, amount: 10, serial: 1 },
```

Comment:
```suggestion
        Bill { owner: User::Alice, amount: 12, serial: 0 },
        Bill { owner: User::Alice, amount: 10, serial: 1 },
```

Reply:
I'm not fully understanding why it shouldn't be this?

Reply:
The serial number cannot be `0` because the initial bill was already with a serial number of `0`. In UTXO model when you spend a bill it gets destroyed and then a new unique serial number is assigned to the outputs.

Reply:
If the transaction does not consume the full bill i.e. the sum of the outputs is actually less than the total input then that just means someone is basically burning the remainder